src/System.Private.CoreLib/shared/System/Text/ASCIIEncoding.cs

```diff
@@ -248,673 +410,338 @@ public override unsafe int GetCharCount(byte[] bytes, int index, int count)
             {
-                charLeftOver = encoder._charLeftOver;
-                fallback = encoder.Fallback as EncoderReplacementFallback;
-
-                // We mustn't have left over fallback data when counting
-                if (encoder.InternalHasFallbackBuffer)
-                {
-                    // We always need the fallback buffer in get bytes so we can flush any remaining ones if necessary
-                    fallbackBuffer = encoder.FallbackBuffer;
-                    if (fallbackBuffer.Remaining > 0 && encoder._throwOnOverflow)
-                        throw new ArgumentException(SR.Format(SR.Argument_EncoderFallbackNotEmpty, this.EncodingName, encoder.Fallback.GetType()));
-
-                    // Set our internal fallback interesting things.
-                    fallbackBuffer.InternalInitialize(charStart, charEnd, encoder, true);
-                }
+                // All elements converted - return immediately.
 
-                Debug.Assert(charLeftOver == 0 || char.IsHighSurrogate(charLeftOver),
-                    "[ASCIIEncoding.GetBytes]leftover character should be high surrogate");
-
-                // Verify that we have no fallbackbuffer, for ASCII its always empty, so just assert
-                Debug.Assert(!encoder._throwOnOverflow || !encoder.InternalHasFallbackBuffer ||
-                    encoder.FallbackBuffer.Remaining == 0,
-                    "[ASCIICodePageEncoding.GetBytes]Expected empty fallback buffer");
+                return numElementsConverted;
             }
             else
             {
-                fallback = this.EncoderFallback as EncoderReplacementFallback;
-            }
-
-
-            // See if we do the fast default or slightly slower fallback
-            if (fallback != null && fallback.MaxCharCount == 1)
-            {
-                // Fast version
-                char cReplacement = fallback.DefaultString[0];
-
-                // Check for replacements in range, otherwise fall back to slow version.
-                if (cReplacement <= (char)0x7f)
-                {
-                    // We should have exactly as many output bytes as input bytes, unless there's a left
-                    // over character, in which case we may need one more.
-                    // If we had a left over character will have to add a ?  (This happens if they had a funky
-                    // fallback last time, but not this time.) (We can't spit any out though
-                    // because with fallback encoder each surrogate is treated as a seperate code point)
-                    if (charLeftOver > 0)
-                    {
-                        // Have to have room
-                        // Throw even if doing no throw version because this is just 1 char,
-                        // so buffer will never be big enough
-                        if (byteCount == 0)
-                            ThrowBytesOverflow(encoder, true);
```

Comment:
>ThrowBytesOverf [](start = 28, length = 15)

are we going to throw the exact Exception with the new change when byteCount == 0?

Reply:
The original code path never should've gotten hit in any real-world application. The only time this code path ever could've run is with the following set of steps:

1. the developer calls `Encoding.ASCII.GetEncoder()`; then
2. sets the fallback buffer to something other than a single ASCII char `EncoderReplacementFallback`; then
3. calls `Encoder.Convert`, passing in a source input that ends with a high surrogate character; then
4. _without clearing the `Encoder` state_, replaces the fallback mechanism with a single ASCII char `EncoderReplacementFallback`; then
5. calls `Encoder.Convert`, passing a zero-length output buffer.

Step (4) above is really kind of sketchy, and I suspect it'd cause other related problems or odd behaviors in practice.

There _are_ some differences with how leftover surrogates are handled. Consider the following code.

```cs
    var encoder = Encoding.ASCII.GetEncoder();

    encoder.Convert("\ud800".ToCharArray(), 0, 1, new byte[0], 0, 0, flush: false, out int charsUsed, out int bytesUsed, out bool completed);

    Console.WriteLine(charsUsed);
    Console.WriteLine(bytesUsed);
    Console.WriteLine(completed);
```

__Before__ this PR, the call to `Convert` throws due to `ThrowBytesOverflow`. __After__ this PR, the call to `Convert` succeeds and returns `(1, 0, false)`. The new behavior is correct because it's entirely possible that the developer could provided an adequately-sized destination buffer on the next call to `Convert`, and because we cannot predict the future we should not fail eagerly.